feat: enzyme db details

e0 opened the pull request:
**Related issue(s) and PR(s)**  
This PR closes #1014 .

<!-- Include below a description of the changes proposed in the pull request -->

**Type of change**  
<!-- Please delete options that are not relevant -->
- [x] New feature (non-breaking change which adds functionality)
 
**List of changes made**  
<!-- Specify what changes have been made and why -->
- Update PostgreSQL schema.
- Add API endpoints for enzymeDB, `enzyme` as well as supplementary data `compound`, `ec`, `reaction`.
- Add page in frontend that displays the entity (`compound`, `ec`, or `reaction`) details as well as listing the related `enzymes`.
- Update the cross references page to accommodate for IDs that are not found in the `neo4j` database.

**Screenshot of the fix**  
<!-- Attach screenshot if relevant -->
<img width="1579" alt="image" src="https://user-images.githubusercontent.com/423498/173382156-0c8a0a59-17e9-44bc-944a-65d5110c774f.png">
<img width="1522" alt="image" src="https://user-images.githubusercontent.com/423498/173381932-429e41fd-50ae-46d0-ad5b-789309794fdc.png">
<img width="637" alt="image" src="https://user-images.githubusercontent.com/423498/173382113-45bd2d65-3ddd-430c-9861-c8715a4c5c54.png">


**Testing**  
<!-- Please delete options that are not relevant -->
- A rebuild is needed due to new dependencies
- Relative urls that can be reused both for production and local testing
- Instructions on how to test

1. Visit an entity's details page. For example:
  - `/enzymedb/compound/C00002`
  - `/enzymedb/ec/2.5.1.18`
  - `/enzymedb/reaction/R00006`
2. Verify that the basic info and cross references (not available for `ec`) tables are displayed.
3. Verify that the enzymes table is displayed.
  - Please try out the filters (must be exact match), table sorting and pagination and verify that they work as expected.
4. Visit an entity's details page that does not exist (for example `/enzymedb/reaction/R000060`) and verify that and error pop-up is shown.

If you run into any problems with the database, please try resetting it. I also noticed that if you don't want any errors during the data import, it helps to delete the extra empty lines at the end of files. For me this was the `abac.txt` and `neq.txt` files.

To reset the database, please do a `git pull` and run the following commands.

```bash

ma-exec pg psql -f scripts/delete-db.sql -U postgres
ma-exec pg psql -f scripts/init-schema.sql -U postgres
ma-exec pg psql -f scripts/seed-data.sql -U postgres

```

**Further comments**  
<!-- Specify questions or related information -->

**Definition of Done checklist**  
- [x] My code meets the Acceptance Criteria
- [x] My code follows the [NBIS style guidelines](https://github.com/NBISweden/development-guidelines)
- [x] I have performed a self-review of my own code and commented any hard-to-understand areas
- [x] Tests and lint/format validations are passing
- [x] My changes generate no new warnings

mihai-sysbio commented:
Since I'm not seeing any explicit mention here, is it intended for this PR to also address [these comments](https://github.com/MetabolicAtlas/MetabolicAtlas/pull/1016#pullrequestreview-1000842668) or is that reserved for another PR? I think that was the plan discussed on Slack but now I'm not sure.

e0 commented:
> Since I'm not seeing any explicit mention here, is it intended for this PR to also address [these comments](https://github.com/MetabolicAtlas/MetabolicAtlas/pull/1016#pullrequestreview-1000842668) or is that reserved for another PR? I think that was the plan discussed on Slack but now I'm not sure.

Thanks for the reminder @mihai-sysbio! I didn't plan on actually changing any files atm regarding those. I should have definitely addressed the comments though.

For the comment https://github.com/MetabolicAtlas/MetabolicAtlas/pull/1016#discussion_r893161495 regarding remote docker compose config, I think it makes more sense to wait with this until the full dataset is provided and tested. Initially I thought maybe there was a chance that we could get the full dataset during the development of this issue. Now that we are wrapping this one up and there is #1023, I think it would make more sense to address it there.

For the comments https://github.com/MetabolicAtlas/MetabolicAtlas/pull/1016#discussion_r893162619 and https://github.com/MetabolicAtlas/MetabolicAtlas/pull/1016#discussion_r893162619 regarding the field types. I have adapted the fields to accommodate the current dataset that's available. We most certainly would need to change some of these later once the full dataset is provided. Please note that it seems that the reference fields, such as `rhea` and `chebi`, have values that are semicolon separated. These should not be capped in length.

For the comment https://github.com/MetabolicAtlas/MetabolicAtlas/pull/1016#discussion_r893164801 regarding database import, I want to remember you mentioning that the full dataset would be provided in a single file as opposed to being divided up like they currently are. If that's the case I think it would make more sense to hard-code a single filename once that's available. If they end up being multiple files then your wildcard suggestion makes a lot of sense.

Overall, I think it's most efficient to wait with these changes until the full dataset is available so that hopefully we can handle them in a single iteration.

mihai-sysbio commented:
Thanks for the detailed response @e0, I am very happy with the suggested strategy.

MalinAhlberg commented:
Great work!
Some initial comments/questions:
- when visiting `/enzymedb/compound/C00003` and order by `ko`, columns move around a bit. Is it possible to make the width of the `Protein` column more fixed, so that the table doesn't change size as much? (maybe just me being annoyed by such things though)

- Kegg id `C00708` is present in eg `/Human-GEM/gem-browser/metabolite/MAM02174c`, but when visiting `/identifier/Kegg/C00708?referenceType=compound` it says "This database identifier is not used by any of the integrated models."

perjo commented on api/src/endpoints/enzymeDb.js:
```diff
@@ -0,0 +1,70 @@
+    }
+  } catch (e) {
+    console.error(e.message);
+    res.sendStatus(400);
```
Seems strange to me to return 400 Bad Request here? The only thing passed in as id and if nothing is found using using that id it should be a 404 and if there are multiple ids for the id the data is bad, not the request? Seems more natural to me to return 500

e0 replied:
Really good point here @perjo. I think we have been a bit sloppy with the error codes. Do you think it's alright if we address this in a new issue? We are using `400` in some other places where changes could be needed too.

<img width="645" alt="image" src="https://user-images.githubusercontent.com/423498/173594676-2f2c322d-6ec0-4b24-931c-349b520e0779.png">

e0 replied:
I think `404` and `500` are really good and sensible suggestions. If we are changing though, I would still like to evaluate more possibilities of code (implementation) changes or alternative status codes. So I would prefer if we can have a bit more discussion about this before changing.

The main concern for me is to do what's convenient for collaborators. In general I would be thinking about the PR size and complexity. In this case since this issue is a blocker for a few others I think that's another good reason to continue this in another issue. Having said that, if the reviewers still think it's better to handle this here, or at least add the changes suggested by Pelle, that would be fine with me as well. Please let me know!

e0 replied:
Apologies again for not double checking earlier. I misunderstood the original situation and thought we only had `400` and that @perjo suggested to change the code to include both `404` and `500`. I took a look now and it seems that I had forgotten that we had already implemented the `404` bit, so since the only suggestion was to change `400` to `500`, that seems trivial enough to be included here.

I have added [a commit](https://github.com/MetabolicAtlas/MetabolicAtlas/pull/1030/commits/9ba0ffa31f8fe7e7d90882a6ed7c6f51cef6a016) for this now. It also includes a change for the `enzymes` endpoint. `404` does not make sense there since if there are 0 matches, it still returns a valid response that looks like `{"enzymes":[],"totalCount":"0"}`.

e0 commented:
I just added a new commit that requires resetting the database. Hopefully this helps solve the sorting problem on numeric values. I also noticed that if you don't want any errors during the data import, it helps to delete the extra empty lines at the end of files. For me this was the `abac.txt` and `neq.txt` files.

To reset the database, please do a `git pull` and run the following commands.

```bash

ma-exec pg psql -f scripts/delete-db.sql -U postgres
ma-exec pg psql -f scripts/init-schema.sql -U postgres
ma-exec pg psql -f scripts/seed-data.sql -U postgres

```

e0 commented:
Just noticed that this latest version is buggy. I set this PR to draft now and will try to fix it before re-requesting. Sorry for the inconvenience.

nanjiangshu reviewed (commented):
Very nice work @e0!

All pages and the API endpoints seem to work nicely when I tested. Below are some of my questions and comments

1. I have to manually create the psql database (with the scripts you provided) after `start-stack`. Is that only I got this problem or it is supposed to be like this since the database handling will be solved in the issue #1023 ? 

2. When filtering the table, I find that it is case sensitive, for example, when filtering the domain with the word `archaea`, the filtered table is empty. I'm not sure if this is ideal as a user. Maybe we should get some feedback from researchers.

3. When filtering the table, only the whole word matching will return some result, if we want to have this feature, I think a drop-down list would be more useful. Otherwise, it might be a bit frustrating when a user tried to filter a column but the table becomes empty once a letter is typed in. My impression is that most of these kind of sort tables filter by partial matching, that is, the size of the table reduces as more letters typed in. 

4. As I already mentioned at the walkthrough meeting, I feel the organism is redundant in the protein name since there is already a column with organism. Shall we create an issue to address this problem?

e0 commented:
Thanks for the feedback @nanjiangshu!

> 1. I have to manually create the psql database (with the scripts you provided) after `start-stack`. Is that only I got this problem or it is supposed to be like this since the database handling will be solved in the issue #1023 ?

This needs to be done manually by everyone right now and I think #1023 is a good place to automate it. Sorry that I didn't make this clear.

___

> 2. When filtering the table, I find that it is case sensitive, for example, when filtering the domain with the word `archaea`, the filtered table is empty. I'm not sure if this is ideal as a user. Maybe we should get some feedback from researchers.

> 3. When filtering the table, only the whole word matching will return some result, if we want to have this feature, I think a drop-down list would be more useful. Otherwise, it might be a bit frustrating when a user tried to filter a column but the table becomes empty once a letter is typed in. My impression is that most of these kind of sort tables filter by partial matching, that is, the size of the table reduces as more letters typed in.

Great points. I think these could be included in one of the search related issues.

___

> 4. As I already mentioned at the walkthrough meeting, I feel the organism is redundant in the protein name since there is already a column with organism. Shall we create an issue to address this problem?

Agree and I think we had already expressed this in the initial discussion a couple of weeks ago. Would be nice to have this removed for the final data output.

mihai-sysbio commented:
> Would be nice to have this removed for the final data output.

The researchers are investigating if the org name can be safely removed from the protein ID.

inghylt reviewed (approved):
Very nice work! To me this is ready to be merged and improvements can be done in another issue

nanjiangshu reviewed (approved):
Looks good to me!
